sound/soc/sof/ipc4-pcm.c

```diff
@@ -462,7 +462,91 @@ static int sof_ipc4_pcm_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm
 	return 0;
 }
 
+static void sof_ipc4_build_time_info(struct snd_sof_dev *sdev, struct snd_sof_pcm_stream *spcm)
+{
+	struct sof_ipc4_copier *host_copier = NULL;
+	struct sof_ipc4_copier *dai_copier = NULL;
+	struct sof_ipc4_llp_reading_slot llp_slot;
+	struct sof_ipc4_timestamp_info *info;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_dai *dai;
+	int i;
+
+	/* find host & dai to locate info in memory window */
+	for_each_dapm_widgets(spcm->list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		if (WIDGET_IS_AIF(swidget->widget->id)) {
+			host_copier = swidget->private;
+		} else if (WIDGET_IS_DAI(swidget->widget->id)) {
+			dai = swidget->private;
+			dai_copier = dai->private;
+		}
+	}
+
+	info = spcm->private;
+	info->host_copier = host_copier;
+	info->dai_copier = dai_copier;
+	info->llp_offset = offsetof(struct sof_ipc4_fw_registers, llp_gpdma_reading_slots) +
+				    sdev->fw_info_box.offset;
+
+	/* find llp slot used by current dai */
+	for (i = 0; i < SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS; i++) {
+		sof_mailbox_read(sdev, info->llp_offset, &llp_slot, sizeof(llp_slot));
+		if (llp_slot.node_id == dai_copier->data.gtw_cfg.node_id)
+			break;
+	}
+
+	if (i == SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS) {
+		/* if no llp gpdma slot is used, check aggregated sdw slot */
+		info->llp_offset = offsetof(struct sof_ipc4_fw_registers, llp_sndw_reading_slots) +
+						sdev->fw_info_box.offset;
+		for (i = 0; i < SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS; i++) {
+			sof_mailbox_read(sdev, info->llp_offset, &llp_slot, sizeof(llp_slot));
+			if (llp_slot.node_id == dai_copier->data.gtw_cfg.node_id)
+				break;
+		}
```

Comment:
same here and btw, you have allocated the `llp_sndw_reading_slots` with a magical `SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS -1`

Reply:
The EVAD gateway is included since it is also supported by SDW

Reply:
So, EVAD is SNDW and it is not SNDW?
Would it need different handling  than SNDW, if not then cover it like SNDW (if it is SNDW), if it needs different handling than you can not just pretend that it is SNDW...

The thing you should avoid is to use different array size in different places based on assumptions.
Ideally the loops should be running on ARRAY_SIZE(the_array_we_are_checking), but since you don't really use the fw_registers struct on real memory...
It helps a lot if the _same_ number is used in definition and when checking the fw_registers.
You see: in the header you are saying that for SNDW we have X-1 number of registers and here you are reading the SNDW registers as there is X number of them.
Sure, you have that EVAD thing in the definition after the SNDW, but that is EVAD, not SNDW.

in one word: confusing.